src/realm_dart_sync.cpp

```diff
@@ -138,3 +137,57 @@ RLM_API void realm_dart_sync_on_subscription_state_changed_callback(realm_userda
         (reinterpret_cast<realm_sync_on_subscription_state_changed_t>(ud->dart_callback))(ud->handle, state);
     });
 }
+
+bool wait_dart_before_continue(realm::util::UniqueFunction<void(realm::util::UniqueFunction<void(bool)>*)>* userCallback)
+{
+    std::condition_variable cv;
+    std::mutex m;
+    bool complete = false;
+    bool success = false;
+
+    realm::util::UniqueFunction<void(bool)> unlockFunc = [&complete, &success, &cv, &m](bool result) {
+        std::unique_lock<std::mutex> lk(m);
+        complete = true;
+        success = result;
+        lk.unlock();
+        cv.notify_one();
+    };
+    (*userCallback)(&unlockFunc);
+
+    std::unique_lock<std::mutex> lk(m);
+    cv.wait(lk, [&complete] {return complete; });
```

Comment:
```suggestion
    realm::util::UniqueFunction unlockFunc = [&](bool result) {
        std::unique_lock lk(m);
        complete = true;
        success = result;
        lk.unlock();
        cv.notify_one();
    };
    (*userCallback)(&unlockFunc);

    std::unique_lock lk(m);
    cv.wait(lk, [&] { return complete; });
```
Let the compiler do the work. This is a "non-escaping" lambda - just use `[&]`